golang/blockchain.go

```diff
@@ -0,0 +1,193 @@
+    }
+
+    newBlock := Block{
+        Index:        int64(len(bc.chain) + 1),
+        Timestamp:    time.Now().UnixNano(),
+        Transactions: bc.transactions,
+        Proof:        proof,
+        PreviousHash: prevHash,
+    }
+
+    bc.transactions = nil
+    bc.chain = append(bc.chain, newBlock)
+    return newBlock
+}
+
+func (bc *Blockchain) NewTransaction(tx Transaction) int64 {
+    bc.transactions = append(bc.transactions, tx)
+    return bc.LastBlock().Index + 1
+}
+
+func (bc *Blockchain) LastBlock() Block {
+    return bc.chain[len(bc.chain)-1]
+}
+
+func (bc *Blockchain) ProofOfWork(lastProof int64) int64 {
+    var proof int64 = 0
+    for !bc.ValidProof(lastProof, proof) {
+        proof += 1
+    }
+    return proof
+}
+
+func (bc *Blockchain) ValidProof(lastProof, proof int64) bool {
+    guess := fmt.Sprintf("%d%d", lastProof, proof)
+    guessHash := ComputeHashSha256([]byte(guess))
+    return guessHash[:4] == "0000"
+}
+
+func (bc *Blockchain) ValidChain(chain *[]Block) bool {
+    lastBlock := (*chain)[0]
+    currentIndex := 1
+    for currentIndex < len(*chain) {
+        block := (*chain)[currentIndex]
+        // Check that the hash of the block is correct
+        if block.PreviousHash != computeHashForBlock(lastBlock) {
+            return false
+        }
+        // Check that the Proof of Work is correct
+        if !bc.ValidProof(lastBlock.Proof, block.Proof) {
+            return false
+        }
+        lastBlock = block
+        currentIndex += 1
+    }
+    return true
+}
+
+func (bc *Blockchain) RegisterNode(address string) bool {
+    u, err := url.Parse(address)
+    if err != nil {
+        return false
+    }
+    return bc.nodes.Add(u.Host)
+}
+
+func (bc *Blockchain) ResolveConflicts() bool {
+    neighbours := bc.nodes
+    newChain := make([]Block, 0)
+
+    // We're only looking for chains longer than ours
+    maxLength := len(bc.chain)
+
+    // Grab and verify the chains from all the nodes in our network
+    for _, node := range neighbours.Keys() {
+        otherBlockchain, err := findExternalChain(node)
+        if err != nil {
+            continue
+        }
+
+        // Check if the length is longer and the chain is valid
+        if otherBlockchain.Length > maxLength && bc.ValidChain(&otherBlockchain.Chain) {
+            maxLength = otherBlockchain.Length
+            newChain = otherBlockchain.Chain
+        }
+    }
+    // Replace our chain if we discovered a new, valid chain longer than ours
+    if len(newChain) > 0 {
+        bc.chain = newChain
+        return true
+    }
+
+    return false
+}
+
+func NewBlockchain() *Blockchain {
+    newBlockchain := &Blockchain{
+        chain:        make([]Block, 0),
+        transactions: make([]Transaction, 0),
+        nodes:        NewStringSet(),
+    }
+    // Initial, sentinel block
+    newBlockchain.NewBlock(100, "1")
+    return newBlockchain
+}
+
+func computeHashForBlock(block Block) string {
+    var buf bytes.Buffer
```

Comment:
For consistency sake, should you consider json-marshal-ing the Block struct, and then calculating the hash?